File: c++/198. House Robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int ls = nums.size(), res = 0, t;
        vector<int> dp(3, 0);
        for (int i = 0; i < ls; i++) {
            t = 0;
            for (int j = 0; j < 2; j++)
                t = max(t, dp[j] + nums[i]);
            res = max(res, t);
            dp.push_back(t);
            dp.erase(dp.begin());
        }
        
        return res;
    }
};

File: go/src/typical_problems/DP/198. House Robber/198. House Robber.go
package main

import "fmt"

func rob(nums []int) int {
	var f1, f2 int
	var f3 int
	for i := range nums {
		f3 = max(f1 + nums[i], f2)
		f1 = f2
		f2 = f3
	}
	return f3
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(rob([]int{1,2,3,1}))
	fmt.Println(rob([]int{2,7,9,3,1}))
}